drivers/mspi/Kconfig

```diff
@@ -0,0 +1,64 @@
+# MSPI driver configuration options
+
+# Copyright (c) 2024 Ambiq Micro Inc. <www.ambiq.com>
+# SPDX-License-Identifier: Apache-2.0
+
+#
+# MSPI Drivers
+#
+menuconfig MSPI
+	bool "Multi-bit Serial Peripheral Interface (MSPI) bus drivers"
+	help
+	  Enable support for the MSPI hardware bus.
+
+if MSPI
+
+config MSPI_ASYNC
+	bool "Asynchronous call support"
+	select POLL
+	help
+	  This option enables the asynchronous API calls.
+
+config MSPI_SLAVE
+	bool "Slave support"
+	help
+	  Enables Driver MSPI slave operations. Slave support depends
+	  on the driver and the hardware it runs on.
+
+config MSPI_INIT_PRIORITY
+	int "Init priority"
+	default 70
+	help
+	  Device driver initialization priority.
+
+config MSPI_COMPLETION_TIMEOUT_TOLERANCE
```

Comment:
Is this intended to be a command timeout parameter? This seems like something we should pass into an API call rather than implementing it at this level- different commands will have different timeout tolerances.

Reply:
I agree, I think I just copied that one from SPI and I haven’t given much thought on that one as the devices I work with does not timeout on command but for the entire transfer. I’m considering add that as a param in `mspi_xfer_t`.  However a maximum timeout for the controller itself may still be meaningful.

Reply:
Personally, I think transfer timeouts make more sense. While most devices will likely have reasonable timeouts, I can imagine something like a chip erase command having a very long timeout needed.

It's also worth noting that in my experience a lot of commands won't have a timeout per se- rather you'd send a command to the device, than poll some type of status bit until the flash indicated the operation was complete.

What types of devices have you encountered where a transfer timeout would be meaningful? I'm just interested to know- not opposed to having the functionality in the API

Reply:
This is the ATXP Series and from what I've seen about Multi-but SPI devices, there are a few that have these requirements.

So in each case, there is a different timing requirement.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/12254017/514aeefa-be77-459f-bad3-9589486752be)

Reply:
Why would the timeout for a specific command not be passed to the MSPI-controller?

Just a question.

Reply:
> Why would the timeout for a specific command not be passed to the MSPI-controller?
> 
> Just a question.

Because MSPI only cares about the transfer, it is all about the transfer. It does not care what happens in between transfers. Unlike SDIO, the protocal dictates a controlled flow and that there is a response for every command. This is not true for SPI. The command timeout is not really the correct term here. I think what you are looking for is the minimum time to wait to start the next transfer. However, the MSPI controller can't be responsible for knowing the internal state of specific devices. This is the responsibility of the specific device driver.

Reply:
> The transfer timeout is mainly for when there is an error occurred. The error could be caused by a misconfiguration or that the device did not receive correct command/addr/data due to timing issue, or even the device did not respond at all due to unknown issue. Surely we wouldn't want to wait forever

I agree, waiting forever does not make sense. However, as you pointed out a lot (if not all) of the device wait times are something the device driver would likely manage. In my experience with MSPI commands, the read phase (where I would expect a timeout to occur) is just handled by the controller sending additional clocks to the device. The *controller* has no idea what the data the device returns here means, only the device driver for the specific attached MSPI device would know this.

What I'm trying to understand is essentially the following- what case exists where we would need the *controller* to return an error due to a timeout? I can absolutely see the controller returning an error due to a transfer failing or not being supported, but these errors will occur at configuration time. I'm trying to understand when the controller itself would timeout. If the device requests a long transfer via the `mspi_transfer` function, the controller should just block until that transfer completes, right?

Reply:
The timeout would also apply to the maximum transfer time as well even if there is no error for the same reason that the upper layer can't wait forever. Thats what the Kconfig `MSPI_COMPLETION_TIMEOUT_TOLERANCE` is for.

Reply:
> Ok, so if this value was set to 200ms and a driver issued a command that (for some reason) took more than 200 ms to transfer the controller would report an error? I guess the idea is to set an upper time bound on the transfer, but wouldn't it make more sense for _each transfer_ to have an upper time bound, not for the framework to have an upper bound set via Kconfig?

There is a timeout parameter in struct mspi_xfer represents the upper bond timeout of a transfer. The Kconfig here is the one for controllers or Hal as I saw timeout macro defined in vendors’ HAL. The idea is the this number should be the biggest timeout of all and the controller should check the timeout in mspi_xfer against this number so that the program wouldn’t wait further after the controller had given up.